### PR TITLE
Polygon Mesh Processing:  Add measure functions not using sqrt

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

Add functions for the case that we only want to compare squared distances (for example for a priority queue).

## Release Management

* Affected package(s): PMP

* Small Feature:  [link ](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/PMP::squared_edge_length)
* Link to compiled documentation (obligatory for small feature) 
   - [`squared_edge_length()`](https://cgal.github.io/5941/v2/Polygon_mesh_processing/group__measure__grp.html#ga30fa03722cd7aa599f6dcb115f54fec5)
   - [`squared_face_area()`](https://cgal.github.io/5941/v2/Polygon_mesh_processing/group__measure__grp.html#ga6eda3738815fd678df225f79ccfc3e03)
* License and copyright ownership: unchanged

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 1]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
 
+/**
+  * \ingroup measure_grp
+  * computes the squared length of an edge of a given polygon mesh.
+  * The edge is given by one of its halfedges, or the edge itself.
+  *
+  * @tparam PolygonMesh a model of `HalfedgeGraph`
+  * @tparam NamedParameters a sequence of \ref bgl_namedparameters "Named Parameters"
+  *
+  * @param h one halfedge of the edge to compute the squared length
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:162]:
```suggestion
  * @param h one halfedge of the edge whose squared length is computed
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 2]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
+ *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
+ *   \cgalParamNEnd
+  * \cgalNamedParamsEnd
+  *
+  * @return the squared length of `h`. The return type `FT` is a number type. It is
+  * either deduced from the `geom_traits` \ref bgl_namedparameters "Named Parameters" if provided,
+  * or the geometric traits class deduced from the point property map
+  * of `pmesh`.
+  *
+  * @sa `edge_length()`
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:187]: `edge_length()` `face_area()` could point to the new functions too.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 3]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
+  */
+template<typename PolygonMesh,
+         typename NamedParameters>
+#ifdef DOXYGEN_RUNNING
+FT
+#else
+typename GetGeomTraits<PolygonMesh, NamedParameters>::type::FT
+#endif
+squared_edge_length(typename boost::graph_traits<PolygonMesh>::halfedge_descriptor h,
+		    const PolygonMesh& pmesh,
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:198]: tabs

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 4]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
+		    const NamedParameters& np)
+{
+  using parameters::choose_parameter;
+  using parameters::get_parameter;
+
+  typename GetVertexPointMap<PolygonMesh, NamedParameters>::const_type
+      vpm = choose_parameter(get_parameter(np, internal_np::vertex_point),
+                             get_const_property_map(CGAL::vertex_point, pmesh));
+
+  return CGAL::squared_distance(get(vpm, source(h, pmesh)),
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:208]: Should grab `gt` from the named parameters and use it.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 5]
```diff
@@ -390,6 +482,87 @@ face_area(typename boost::graph_traits<TriangleMesh>::face_descriptor f,
   return face_area(f, tmesh, CGAL::Polygon_mesh_processing::parameters::all_default());
 }
 
+/**
+  * \ingroup measure_grp
+  * computes the squared area of a face of a given
+  * triangulated surface mesh.
+  *
+  * @tparam TriangleMesh a model of `HalfedgeGraph`
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:490]: FaceGraph, needs to be fixed in the other functions too.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 6]
```diff
@@ -390,6 +482,87 @@ face_area(typename boost::graph_traits<TriangleMesh>::face_descriptor f,
...
+  * @tparam NamedParameters a sequence of \ref bgl_namedparameters "Named Parameters"
+  *
+  * @param f the face of which the squared area is computed
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:493]:
```suggestion
  * @param f the face whose squared area is computed
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 7]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
+  *   \cgalParamNEnd
+  *
+ *   \cgalParamNBegin{geom_traits}
+ *     \cgalParamDescription{an instance of a geometric traits class}
+ *     \cgalParamType{a class model of `Kernel`}
+ *     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
+ *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
+ *   \cgalParamNEnd
+  * \cgalNamedParamsEnd
+  *
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:181]: \pre h non-null ?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 8]
```diff
@@ -150,6 +151,93 @@ edge_length(typename boost::graph_traits<PolygonMesh>::edge_descriptor e,
...
+#else
+typename GetGeomTraits<PolygonMesh, NamedParameters>::type::FT
+#endif
+squared_edge_length(typename boost::graph_traits<PolygonMesh>::halfedge_descriptor h,
+		    const PolygonMesh& pmesh,
+		    const NamedParameters& np)
+{
+  using parameters::choose_parameter;
+  using parameters::get_parameter;
+
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:203]: Could assert that h is non null

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h [hunk 9]
```diff
@@ -390,6 +482,87 @@ face_area(typename boost::graph_traits<TriangleMesh>::face_descriptor f,
...
+  typename GetVertexPointMap<TriangleMesh, CGAL_PMP_NP_CLASS>::const_type
+      vpm = choose_parameter(get_parameter(np, internal_np::vertex_point),
+                             get_const_property_map(CGAL::vertex_point, tmesh));
+
+  halfedge_descriptor hd = halfedge(f, tmesh);
+  halfedge_descriptor nhd = next(hd, tmesh);
+
+  typedef typename GetGeomTraits<TriangleMesh, CGAL_PMP_NP_CLASS>::type GT;
+  GT traits = choose_parameter<GT>(get_parameter(np, internal_np::geom_traits));
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:549]:
```suggestion
  typename GetVertexPointMap<TriangleMesh, CGAL_PMP_NP_CLASS>::const_type
      vpm = choose_parameter(get_parameter(np, internal_np::vertex_point),
                             get_const_property_map(CGAL::vertex_point, tmesh));

  typedef typename GetGeomTraits<TriangleMesh, CGAL_PMP_NP_CLASS>::type GT;
  GT traits = choose_parameter<GT>(get_parameter(np, internal_np::geom_traits));

  halfedge_descriptor hd = halfedge(f, tmesh);
  halfedge_descriptor nhd = next(hd, tmesh);
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 1 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:162]: I changed it, also for `edge_length()`

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 7 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:181]: Is n't that in general the case?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 2 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:187]: done.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 7 above)
Comment (MaelRL) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:181]: You made the requirement explicit in the doc of `squared_face_area()` (or at least it was so in `face_area()`, I guess). It's just symmetry, not very important; the precondition in the actual code is more important.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 8 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:203]: Same question as for the preconditon.   @sloriot what is your opinion on this.   As @janetournois me just explaind there is an assertion for `face_area()`  and not for `edge_length()`  as it is "easier"  to have a null face, namely by writing `face_area(face(he, mesh), mesh)`  than having a null edge.

Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:203]: I agree with Jane but it does not really hurt

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 6 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:493]: changed globally

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 7 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:184]: I added an assertion.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h (hunk 8 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/measure.h:206]: I added an assertion.

**Comment by lrineau** [comment]
/testme

(I want to check if that action still work after all the changes I made to our credentials.)

**Comment by lrineau** [comment]
> The testsuite is lauched. results will be found, after it is done, here : [cgal.geometryfactory.com/~cgaltest/test_suite/TESTRESULTS/index.shtml](https://cgal.geometryfactory.com/~cgaltest/test_suite/TESTRESULTS/index.shtml)

Yeah, that worked! See the results at https://cgal.geometryfactory.com/~cgaltest/test_suite/TESTRESULTS/results-5.4-Ic-5941.shtml.

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.3-Ic-53

**Comment by lrineau** [comment]
This small feature does not add anything to `CHANGES.md`. Is that on purpose?

**Pull request merged**